File: examples/serial/src/Provider.fs
module SerialProvider.Provider

// The serial device provider (Plan 16, part E): an ordinary MCP server that happens to own
// some ttys.
//
// Nothing about it is Yession-shaped, which is the whole point — it is separately shippable
// and useful to any MCP client. It exists because `/dev/ttyACM0` cannot speak MCP: something
// has to own the OS-level port claim, enumerate what is attached, and hand out an address a
// byte stream can be opened on. A device that speaks MCP natively needs none of this and is
// declared to a session directly.
//
// Two legs, as Plan 16 split them and Plan 17 wired up:
//
//   control  MCP over Streamable HTTP at /mcp   — list, acquire, configure, release
//   data     a WebSocket at /attach/<token>     — the bytes, both ways
//
// joined ONE WAY by the attach ticket: `acquire_device` answers with a url, and the client
// opens it. Nothing flows back the other way, so a terminal closing does not disturb the
// control leg and the control leg dropping does not close a terminal already streaming.

open System
open System.Collections.Generic
open SerialProvider
open SerialProvider.Interop

/// A device that has been claimed, and the port behind it once somebody attaches.
type private Claimed =
    { Device : SerialDevice
      /// The MCP session that took it. A claim dies with the session that took it, which is
      /// what stops a crashed client holding hardware forever.
      Holder : string
      mutable Settings : SerialSettings
      /// The single-use token in the attach url. Single-use because it IS the authority to
      /// reach the device: a token that could be replayed would let a second client take a
      /// stream the claim says is exclusive.
      mutable Token : string
      /// The live port, once the data leg opened. `None` while the claim is held but nobody
      /// has attached — an acquire that is never followed by an attach holds the DEVICE, not
      /// a file descriptor.
      mutable Port : Ports.OpenPort option }

/// The provider, as its host process sees it.
type SerialProvider =
    { /// Handle one control-leg request. `true` when it claimed the path.
      TryHandle : IncomingMessage -> ServerResponse -> bool
      /// Attach WebSocket upgrades to an HTTP server. Separate from `TryHandle` because an
      /// upgrade is not a request — it is an event on the server object.
      Serve : HttpServer -> unit
      /// What is claimed right now (observability, and what the tests read).
      Claims : unit -> SerialClaim list }

// --- the tools, declared ------------------------------------------------------------------
//
// One list, read by BOTH `tools/list` and `tools/call`, so the two cannot disagree about what
// this server accepts. The descriptions are written for the MODEL, not for a person reading
// the source: they say what the tool is for, what it costs, and what to do next, because a
// tool description is the only documentation the caller will ever see.

let Namespace = "serial"

let private deviceArg = Mcp.Schema.required "device_id" "string" "the device_id list_devices returned"

let tools : Mcp.Tool list =
    [ { Name = "list_devices"
        Title = None
        Description =
            "Every serial device attached to this host that this provider recognises, with a stable device_id, its make and model, and whether it is free. Unrecognised ports are deliberately not listed — a serial port that is not a known device is usually the machine's own console. Call this again after somebody plugs something in; the list is read from the OS each time."
        InputSchema = Mcp.Schema.ofFields [] }
      { Name = "acquire_device"
        Title = Some "Acquire a serial device"
        Description =
            "Claim a device for exclusive use and get back a URL to open its byte stream on. One holder at a time, because the OS gives out one handle: if somebody else has it, this says who. Release it when you are done."
        InputSchema = Mcp.Schema.ofFields [ deviceArg ] }
      { Name = "configure_device"
        Title = None
        Description =
            "Set the line parameters of a device you hold. Takes effect on the NEXT attach, so configure before you open the stream. Defaults are 115200 8N1, which is what most boards boot at."
        InputSchema =
            Mcp.Schema.ofFields
                [ deviceArg
                  Mcp.Schema.required "baud_rate" "integer" "e.g. 9600, 115200"
                  Mcp.Schema.optional "data_bits" "integer" "5, 6, 7 or 8 (default 8)"
                  Mcp.Schema.optional "stop_bits" "integer" "1 or 2 (default 1)"
                  Mcp.Schema.optional "parity" "string" "none, even or odd (default none)" ] }
      { Name = "release_device"
        Title = None
        Description =
            "Give a device back, closing its stream if one is open. Safe to call on a device you do not hold."
        InputSchema = Mcp.Schema.ofFields [ deviceArg ] } ]

// --- the answers ---------------------------------------------------------------------------

/// Build a provider over an engine. `origin` is the address a CLIENT should use to reach the
/// data leg — passed in rather than derived from a request, because the provider may sit
/// behind something and only its operator knows.
let create (engine: Ports.SerialEngine) (origin: unit -> string) (version: string) : SerialProvider =
    // device_id -> claim. A `Dictionary` rather than a map in a ref because a claim is
    // mutated by three different legs (configure, attach, the port closing) and threading a
    // whole map through each would be bookkeeping with no reader.
    let claimed = Dictionary<string, Claimed> ()
    // token -> device_id, so an attach resolves in one lookup and a spent token is removed
    // rather than searched for.
    let tokens = Dictionary<string, string> ()

    let devices () : Async<SerialDevice list> =
        async {
            match! engine.List () with
            | Error _ -> return []
            | Ok ports -> return Discovery.devices ports
        }

    let mintToken () = randomId ()

    let release (deviceId: string) : unit =
        match claimed.TryGetValue deviceId with
        | false, _ -> ()
        | true, claim ->
            claim.Port |> Option.iter (fun port -> port.Close ())
            claim.Port <- None
            tokens.Remove claim.Token |> ignore
            claimed.Remove deviceId |> ignore

    // --- tool bodies -------------------------------------------------------------------
    //
    // Each answers TEXT, because that is what a model reads, and each says which state it is
    // in rather than only whether it worked. `acquire_device` in particular has to name the
    // holder when it refuses: a bare "unavailable" reads as a hardware fault.

    let listDevices () : Async<string> =
        async {
            let! found = devices ()
            if List.isEmpty found then
                return
                    "no serial devices attached (or none this provider recognises — unrecognised ports are not listed)"
            else
                return
                    found
                    |> List.map (fun device ->
                        let state =
                            match claimed.TryGetValue device.DeviceId with
                            | true, claim -> sprintf "held by %s" claim.Holder
                            | _ -> "free"
                        sprintf
                            "%s — %s %s at %s (%s)"
                            device.DeviceId
                            device.Make
                            device.Model
                            device.Port.Path
                            state)
                    |> String.concat "\n"
        }

    let acquire (holder: string) (deviceId: string) : Async<string> =
        async {
            let! found = devices ()
            match found |> List.tryFind (fun d -> d.DeviceId = deviceId) with
            | None -> return sprintf "no device '%s' is attached — call list_devices" deviceId
            | Some device ->
                match claimed.TryGetValue deviceId with
                | true, claim when claim.Holder <> holder ->
                    // Naming the holder is the whole point of this branch.
                    return sprintf "%s is already held by %s — it is in use, not broken" deviceId claim.Holder
                | true, claim ->
                    return
                        sprintf
                            "you already hold %s — attach it at %s/attach/%s"
                            deviceId
                            (origin ())
                            claim.Token
                | _ ->
                    let token = mintToken ()
                    claimed.[deviceId] <-
                        { Device = device
                          Holder = holder
                          Settings = SerialSettings.defaults
                          Token = token
                          Port = None }
                    tokens.[token] <- deviceId
                    return
                        sprintf
                            "%s is yours. Attach its byte stream at %s/attach/%s — the stream is live-only, so it has no exit codes and nothing there reports whether a command succeeded."
                            deviceId
                            (origin ())
                            token
        }

    let configure (holder: string) (deviceId: string) (settings: SerialSettings) : Async<string> =
        async {
            match claimed.TryGetValue deviceId with
            | false, _ -> return sprintf "you do not hold %s — acquire it first" deviceId
            | true, claim when claim.Holder <> holder ->
                return sprintf "%s is held by %s" deviceId claim.Holder
            | true, claim ->
                match SerialSettings.validate settings with
                | Error e -> return sprintf "those line settings are not usable: %s" e
                | Ok settings ->
                    claim.Settings <- settings
                    // Stated rather than silently true: a caller who configures a device
                    // whose stream is already open and sees no change would reasonably
                    // conclude the call did nothing.
                    let already = if claim.Port.IsSome then " Re-attach for it to take effect." else ""
                    return
                        sprintf
                            "%s is set to %d %d%s%d.%s"
                            deviceId
                            settings.BaudRate
                            settings.DataBits
                            (match settings.Parity with
                             | "none" -> "N"
                             | "even" -> "E"
                             | "odd" -> "O"
                             | other -> (other.Substring (0, 1)).ToUpperInvariant ())
                            settings.StopBits
                            already
        }

    let releaseTool (holder: string) (deviceId: string) : Async<string> =
        async {
            match claimed.TryGetValue deviceId with
            | false, _ -> return sprintf "%s was not held" deviceId
            | true, claim when claim.Holder <> holder ->
                return sprintf "%s is held by %s, not by you" deviceId claim.Holder
            | true, _ ->
                release deviceId
                return sprintf "released %s" deviceId
        }

    let invoke (holder: string) (call: Mcp.ToolCall) : Async<Result<string, string>> =
        let ok (body: Async<string>) = async { let! text = body in return Ok text }
        let deviceOf (json: string) =
            match SerialWire.deviceId json with
            | Error e -> Error e
            | Ok id -> Ok id
        async {
            match call.Name with
            | "list_devices" -> return! ok (listDevices ())
            | "acquire_device" ->
                match deviceOf call.Arguments with
                | Error e -> return Error e
                | Ok id -> return! ok (acquire holder id)
            | "release_device" ->
                match deviceOf call.Arguments with
                | Error e -> return Error e
                | Ok id -> return! ok (releaseTool holder id)
            | "configure_device" ->
                match SerialWire.configure call.Arguments with
                | Error e -> return Error e
                | Ok (id, settings) -> return! ok (configure holder id settings)
            | other -> return Error (sprintf "no tool '%s'" other)
        }

    // --- the control leg ----------------------------------------------------------------

    let mcp = Mcp.create { Name = "serial"; Version = version; Tools = tools; Invoke = invoke }

    // --- the data leg -------------------------------------------------------------------

    let attach (token: string) (peer: Ws.WsPeer) : Ws.WsHandlers =
        match tokens.TryGetValue token with
        | false, _ ->
            peer.Control (SerialWire.exited 1)
            peer.Close ()
            { Data = ignore; Control = ignore; Closed = ignore }
        | true, deviceId ->
            let claim = claimed.[deviceId]
            // SPENT on use. The token is the authority to reach the device, and one that
            // could be replayed would hand a second client the stream the claim says is
            // exclusive.
            tokens.Remove token |> ignore
            let mutable port : Ports.OpenPort option = None
            let finish (why: string) =
                // The in-band termination frame, which is what tells a client the difference
                // between "the device went away" and "the network did". An abrupt close
                // carries nothing, which is exactly why this exists.
                peer.Control (SerialWire.exited 0)
                ignore why
                peer.Close ()
            Async.StartImmediate (
                async {
                    match! engine.Open claim.Device.Port.Path claim.Settings peer.Send finish with
                    | Error reason ->
                        peer.Send (sprintf "could not open %s: %s\r\n" claim.Device.Port.Path reason)
                        finish reason
                    | Ok opened ->
                        port <- Some opened
                        claim.Port <- Some opened
                })
            { Data = fun text -> port |> Option.iter (fun p -> p.Write text)
              // A serial line has no size and no signals: `resize` is nothing to do, and
              // `kill` closes the stream rather than killing anything. Saying so here beats
              // a client discovering it by silence.
              Control =
                fun text ->
                    match SerialWire.control text with
                    | SerialWire.Kill ->
                        port |> Option.iter (fun p -> p.Close ())
                        claim.Port <- None
                        finish "closed by the client"
                    | SerialWire.Resize
                    | SerialWire.Unknown -> ()
              Closed =
                fun () ->
                    port |> Option.iter (fun p -> p.Close ())
                    claim.Port <- None }

    { TryHandle = mcp.TryHandle
      Serve =
        fun server ->
            Ws.serve server (fun path ->
                if path.StartsWith "/attach/" then Some (attach (path.Substring "/attach/".Length))
                else None)
      Claims =
        fun () ->
            claimed
            |> Seq.map (fun pair ->
                { DeviceId = pair.Key
                  Holder = pair.Value.Holder
                  Settings = pair.Value.Settings
                  AttachToken = pair.Value.Token })
            |> Seq.sortBy (fun claim -> claim.DeviceId)
            |> List.ofSeq }

File: examples/jumpstarter/src/jumpstarter_provider/provider.py
"""The provider: eight tools, one claim, and the MCP server they hang off.

Nothing here is Yession-shaped. Point any MCP client at the control url and it works, which
is the whole point of an out-of-process provider and the reason this is an example rather
than a built-in.

The claim is the interesting part. Jumpstarter arbitrates access with LEASES, and leases
live in the controller — which a direct-mode exporter does not have. So the arbitration is
this server's, and it is one claim over the WHOLE exporter rather than one per driver:
powering the board off invalidates the console session, so a claim per driver would be a
promise of isolation the hardware does not keep.
"""

from __future__ import annotations

import time
from contextlib import contextmanager
from dataclasses import dataclass
from typing import Any, Callable, Iterator

import anyio
from mcp.server.mcpserver import Context, MCPServer
from mcp.types import CallToolResult, TextContent

from .exporter import Exporter, ExporterError
from .stream import Console

# A claim follows a live MCP session. A session is alive while it is TALKING: every request
# under its id — a tool call, or the `tools/list` a client polls with — pushes the deadline
# out, and a client that has gone silent for this long has gone.
#
# ...or while it is LISTENING, which is the same fact arriving differently. Streamable HTTP
# lets a client open one GET and leave it open for the session's whole life; measured by
# requests that client spoke once and vanished, while it is in fact the most connected a
# client gets. `Claims.streaming` is that second reading, and it is here rather than in the
# caller because a rule that holds only while the caller keeps polling is not the claim's
# rule — it is the caller's habit, and the next client has not got it.
#
# One rule, not two: an explicit goodbye (`release`, or the DELETE that ends the session)
# and a crashed client are the same event seen at different speeds, and a second mechanism
# for the first case would be a second answer to "who holds this".
DEFAULT_CLAIM_TTL_SECONDS = 300.0

# Where the data leg lives, on the same listener as `/mcp`.
ATTACH_PATH = "/attach/"


@dataclass
class _Claim:
    holder: str
    taken_at: float
    last_seen: float


class Claims:
    """Who holds the exporter, and for how much longer.

    Its clock is a parameter so a test can age a claim without sleeping for five minutes.
    """

    def __init__(
        self,
        ttl_seconds: float = DEFAULT_CLAIM_TTL_SECONDS,
        now: Callable[[], float] = time.monotonic,
        on_release: Callable[[], None] = lambda: None,
    ) -> None:
        self.ttl_seconds = ttl_seconds
        self._now = now
        self._on_release = on_release
        self._claim: _Claim | None = None
        # Counted, not flagged: a client may hold more than one open stream at a time, and
        # the first to close must not answer for the rest.
        self._streaming: dict[str, int] = {}

    def touch(self, holder: str) -> None:
        """A session said something. Called for EVERY request, including the ones the MCP
        server answers itself — which is why it lives at the HTTP layer rather than in a
        tool body."""
        self.expire()
        if self._claim is not None and self._claim.holder == holder:
            self._claim.last_seen = self._now()

    @contextmanager
    def streaming(self, holder: str) -> Iterator[None]:
        """For as long as this session holds a stream open, it is talking.

        Entered around the request that IS the stream, so the claim follows that request's
        lifetime rather than its arrival — the one case the deadline cannot measure, because
        nothing further arrives to measure.
        """
        self._streaming[holder] = self._streaming.get(holder, 0) + 1
        try:
            yield
        finally:
            remaining = self._streaming.get(holder, 0) - 1
            if remaining > 0:
                self._streaming[holder] = remaining
            else:
                self._streaming.pop(holder, None)
                # The deadline starts NOW rather than from whenever this session last sent
                # a request. It has been alive the whole time the stream was open, so
                # measuring its silence from before that would expire a client that has
                # said nothing wrong — instantly, if the stream outlived one TTL.
                claim = self._claim
                if claim is not None and claim.holder == holder:
                    claim.last_seen = self._now()

    def expire(self) -> None:
        claim = self._claim
        if claim is None:
            return
        if self._streaming.get(claim.holder):
            return
        if self._now() - claim.last_seen > self.ttl_seconds:
            self._claim = None
            self._on_release()

    @property
    def holder(self) -> str | None:
        self.expire()
        return self._claim.holder if self._claim else None

    def acquire(self, holder: str) -> tuple[bool, str | None]:
        """(took it, who has it instead)."""
        self.expire()
        if self._claim is None:
            now = self._now()
            self._claim = _Claim(holder=holder, taken_at=now, last_seen=now)
            return True, None
        if self._claim.holder == holder:
            return True, None
        return False, self._claim.holder

    def release(self, holder: str) -> bool:
        self.expire()
        if self._claim is None or self._claim.holder != holder:
            return False
        self._claim = None
        self._on_release()
        return True


def _said(text: str) -> CallToolResult:
    """Prose, as the one shape a tool that SOMETIMES carries `_meta` may answer with. The
    SDK refuses a union return, and rightly: a signature that is sometimes a result and
    sometimes a string is two contracts."""
    return CallToolResult(content=[TextContent(type="text", text=text)])


def _short(holder: str) -> str:
    """An MCP session id is 32 hex characters, and a refusal that quotes all of it reads as
    a hash rather than as somebody. The first eight are enough to tell two clients apart."""
    return holder[:8]


class Provider:
    """The MCP server, the exporter behind it, and the claim between them."""

    def __init__(
        self,
        exporter: Exporter,
        claims: Claims,
        version: str,
        origin: Callable[[], str],
        console: Console,
    ) -> None:
        self.exporter = exporter
        self.claims = claims
        # Where this server's own data leg is, resolved LATE: a provider bound to port 0
        # does not know its address until it is listening, and every test binds port 0.
        self.origin = origin
        self.console = console
        # `jumpstarter` is the namespace every tool of this server lands in for a client
        # that prefixes them (Yession shows `mcp__jumpstarter__power`), so it is not a
        # label — it is half of every name the model sees.
        self.mcp = MCPServer("jumpstarter", version=version)
        self._register()

    # --- the claim, as the tools see it ------------------------------------------------

    def _holder(self, ctx: Context) -> str:
        # The session id is on the request, which is what makes a claim outlive a tool call
        # without the caller having to carry a token around.
        return dict(ctx.headers).get("mcp-session-id", "")

    def _refusal(self, ctx: Context) -> str | None:
        """Why this caller may not touch the hardware, in the words it should read.

        A refusal is an ANSWER, not an error: "somebody else has it" is a fact about the
        world, and a caller told it as a failure reasonably concludes something is broken
        and tries to fix it.
        """
        holder = self._holder(ctx)
        if not holder:
            return "this request carries no MCP session id, so nothing can be claimed by it"
        current = self.claims.holder
        if current is None:
            return "nobody holds this exporter — call acquire first"
        if current != holder:
            return f"the exporter is held by {_short(current)}, not by you — it is in use, not broken"
        return None

    # --- the tools ---------------------------------------------------------------------
    #
    # The descriptions are written for the MODEL, not for a person reading the source: what
    # the tool is for, what it costs, and what to do next. A tool description is the only
    # documentation its caller will ever see.

    def _register(self) -> None:
        mcp = self.mcp

        @mcp.tool()
        async def status(ctx: Context) -> str:
            """What this exporter is: whether it can be reached, which drivers it exports (power, serial, storage and so on), and who currently holds it. Costs nothing and needs no claim — call it first, and call it again if something stops answering."""
            holder = self.claims.holder
            if holder is None:
                held = "nobody holds it"
            elif holder == dict(ctx.headers).get("mcp-session-id", ""):
                held = "you hold it"
            else:
                held = f"held by {_short(holder)} — it is in use, not broken"
            try:
                drivers = await _off_loop(self.exporter.drivers)
            except ExporterError as error:
                return f"the exporter is not answering: {error}"
            listed = "\n".join(f"  {line}" for line in drivers) or "  (it exports nothing)"
            return f"exporter at {self.exporter.host}, {held}.\nDrivers:\n{listed}"

        @mcp.tool()
        async def acquire(ctx: Context) -> CallToolResult:
            """Claim this exporter for your exclusive use, and get its console as a stream. Everything that touches the hardware needs the claim, because powering the board off cuts the console out from under anyone else. One holder at a time: if somebody else has it, this says who. Release it when you are done."""
            holder = self._holder(ctx)
            if not holder:
                return _said("this request carries no MCP session id, so nothing can be claimed by it")
            took, other = self.claims.acquire(holder)
            if not took:
                return _said(f"the exporter is already held by {_short(other or '')} — it is in use, not broken")
            held = (
                "the exporter is yours. It stays yours while your session keeps talking to this "
                "server or keeps its event stream open, and is released automatically after "
                f"{int(self.claims.ttl_seconds)}s of neither."
            )
            # The console, as a stream a CLIENT can open — in `_meta`, which is the place MCP
            # reserves for data meant for the client rather than the model. A client that has
            # never heard of the key ignores it and still gets the prose, and still has the
            # three console tools; one that knows it turns the console into a terminal people
            # can watch and type into.
            offer = self.console.offer(self.origin(), holder)
            if offer is None:
                return _said(f"{held} Its console is already attached to a terminal.")
            return CallToolResult(
                content=[TextContent(type="text", text=held)],
                meta={"dev.yession/stream": offer},
            )

        @mcp.tool()
        async def release(ctx: Context) -> str:
            """Give the exporter back, closing the console if one is open. Safe to call when you do not hold it."""
            holder = self._holder(ctx)
            if not self.claims.release(holder):
                current = self.claims.holder
                if current is None:
                    return "nobody held the exporter"
                return f"the exporter is held by {_short(current)}, not by you"
            return "released the exporter"

        @mcp.tool()
        async def power(ctx: Context, action: str) -> str:
            """Power the device under test on, off, or cycle it (off, a pause, then on) — the usual way to recover a board that has stopped responding. Needs the claim. Anything on the console at the time is lost; re-read it after. There is no way to ASK whether the board is on: jumpstarter's power drivers take commands and report measurements, so read the volts and amps with driver_call power read instead."""
            refusal = self._refusal(ctx)
            if refusal:
                return refusal
            wanted = action.strip().lower()
            if wanted not in ("on", "off", "cycle"):
                return f"'{action}' is not a power action — use on, off or cycle"
            try:
                if wanted == "cycle":
                    await _off_loop(lambda: self.exporter.call("power", "off", []))
                    await anyio.sleep(2.0)
                    await _off_loop(lambda: self.exporter.call("power", "on", []))
                else:
                    await _off_loop(lambda: self.exporter.call("power", wanted, []))
            except ExporterError as error:
                return f"the power driver refused: {error}"
            return f"power {wanted} done"

        @mcp.tool()
        async def driver_call(ctx: Context, driver: str, method: str, args: list[Any] | None = None) -> str:
            """Call any method on any driver this exporter exports — the general case behind the power and console tools. `driver` is a name from status (dotted for a nested one, like "bench.power"), `method` is one of its methods, `args` are positional. Call with no method to be told exactly what that driver offers, which is the reliable way to find out — a method that streams (like power read) is drained to its first few items and says so. Needs the claim."""
            refusal = self._refusal(ctx)
            if refusal:
                return refusal
            try:
                if not method:
                    offered = await _off_loop(lambda: self.exporter.methods(driver))
                    return f"{driver} offers: {', '.join(offered)}"
                result = await _off_loop(lambda: self.exporter.call(driver, method, list(args or [])))
            except ExporterError as error:
                return str(error)
            # `!r` is the exporter's job now, and it already did it: `call` answers with a
            # STRING it shaped (a repr, or a drained stream). Repr-ing that again wrapped
            # every answer in quotes, so a list of readings read as one long string.
            return f"{driver}.{method} returned {result}"

        @mcp.tool()
        async def serial_send(ctx: Context, data: str) -> str:
            """Write to the device's serial console. Nothing is appended, so end a command with \\n yourself. The console opens on first use and stays open until you release the exporter. Needs the claim. This says only that the bytes went out — read or expect to find out what happened."""
            refusal = self._refusal(ctx)
            if refusal:
                return refusal
            # One writer, and while a terminal is attached it is the terminal's — where a
            # lease says who is typing and everyone can see the keystrokes. Writing here as
            # well would be a second door onto one console, past that lease.
            if self.console.attached:
                return (
                    "the console is attached to a terminal — type into it there "
                    "(write_terminal), where people can see who is typing"
                )
            try:
                await _off_loop(lambda: self.exporter.console_send(data))
            except ExporterError as error:
                return f"the console refused the write: {error}"
            return f"sent {len(data)} characters"

        @mcp.tool()
        async def serial_read(ctx: Context, timeout_seconds: float = 2.0) -> str:
            """Everything the console has said since you last read it, up to a pause of timeout_seconds. Needs the claim. Use expect instead when you know what you are waiting for — this one always waits the whole timeout out."""
            refusal = self._refusal(ctx)
            if refusal:
                return refusal
            try:
                seen = await self.console.read(_timeout(timeout_seconds))
            except ExporterError as error:
                return f"the console is not readable: {error}"
            return seen if seen else "(the console said nothing)"

        @mcp.tool()
        async def serial_expect(ctx: Context, pattern: str, timeout_seconds: float = 10.0) -> str:
            """Wait for the console to say something matching pattern (a Python regular expression) and return everything up to and including it — a login prompt, a boot banner, a shell prompt after a command. Returns as soon as it matches. On a timeout you still get what WAS said, which is usually where the answer is. Needs the claim."""
            refusal = self._refusal(ctx)
            if refusal:
                return refusal
            try:
                matched, seen = await self.console.expect(pattern, _timeout(timeout_seconds))
            except ExporterError as error:
                return f"the console is not readable: {error}"
            if matched:
                return seen
            body = seen if seen else "(nothing at all)"
            return f"'{pattern}' did not appear within {_timeout(timeout_seconds):.0f}s. The console said:\n{body}"

    # --- the app -----------------------------------------------------------------------

    def app(self) -> Any:
        """The ASGI app, with the one thing the MCP server does not do for us wrapped
        around it: seeing every request, so a claim can follow a session it never hears
        about at the tool layer (`tools/list` is answered by the server itself, and a
        polling client is a live client — as is a streaming one, below)."""
        inner = self.mcp.streamable_http_app()
        claims = self.claims
        console = self.console

        async def app(scope: Any, receive: Any, send: Any) -> None:
            if scope["type"] == "http":
                session = _session_of(scope)
                if session:
                    claims.touch(session)
                else:
                    claims.expire()
                # The optional server->client stream is a GET that arrives once and stays
                # open for the session's life. Touching it on arrival would credit it with
                # one instant of liveness and then let the claim expire underneath a client
                # that is still connected, so the claim follows the stream while it is OPEN.
                if session and scope.get("method") == "GET":
                    with claims.streaming(session):
                        await inner(scope, receive, send)
                    return
            # The data leg (Plan 19), on the same port as the control leg — which is the
            # whole reason a WebSocket was the right shape: the upgrade rides the listener
            # the provider already has.
            if scope["type"] == "websocket":
                path = scope.get("path", "")
                if path.startswith(ATTACH_PATH):
                    await console.serve(path[len(ATTACH_PATH) :], receive, send)
                    return
                await send({"type": "websocket.close", "code": 1008})
                return
            await inner(scope, receive, send)

        return app


def _session_of(scope: Any) -> str:
    """The MCP session id off an ASGI request, or "" for a request that carries none."""
    for name, value in scope.get("headers", []):
        if name == b"mcp-session-id":
            return value.decode()
    return ""


def _timeout(seconds: float) -> float:
    """A timeout is a promise to answer, so it is bounded: a caller that asks for an hour
    would hold the whole server on one request."""
    return max(0.1, min(float(seconds), 120.0))


async def _off_loop(work: Callable[[], Any]) -> Any:
    """Everything the exporter does is a blocking queue round-trip, and doing it on the
    event loop would stall every other session for the length of a console read."""
    return await anyio.to_thread.run_sync(work)


def create(
    host: str,
    console: str,
    ttl_seconds: float,
    version: str,
    origin: Callable[[], str] = lambda: "",
) -> Provider:
    exporter = Exporter(host=host, console_name=console)
    stream = Console(exporter=exporter, off_loop=_off_loop)

    # A claim ending closes the console, wherever that claim ended: `release`, or silence.
    # The console is the DUT's, not the holder's, and leaving it open would hand the next
    # holder a stream mid-sentence. The STREAM is the claim's for the same reason, so its
    # token dies with it.
    def released() -> None:
        stream.release()
        exporter.close_console_quietly()

    claims = Claims(ttl_seconds=ttl_seconds, on_release=released)
    return Provider(exporter=exporter, claims=claims, version=version, origin=origin, console=stream)
